src/ceph-volume/ceph_volume/tests/util/test_device.py

```diff
@@ -218,7 +218,7 @@ def test_reject_not_acceptable_device(self, device_info):
         assert not disk.available
 
     def test_reject_readonly_device(self, device_info):
-        data = {"/dev/cdrom": {"ro": 1}}
+        data = {"/dev/sr0": {"ro": 1}}
```

Comment:
> > @guits Does `ceph-volume tox testing` is conducted on VM/bare-metal with actual cd-rom?
> 
> ```
> [jenkins-build@irvingi01 ~]$ ls -l /dev/cdrom
> ls: cannot access /dev/cdrom: No such file or directory
> ```

If so then this change is not necessary but previously this was the issue with ceph-volume tox test. Does every VM for test is the same?

Reply:
> > > @guits Does `ceph-volume tox testing` is conducted on VM/bare-metal with actual cd-rom?
> > 
> > 
> > ```
> > [jenkins-build@irvingi01 ~]$ ls -l /dev/cdrom
> > ls: cannot access /dev/cdrom: No such file or directory
> > ```
> 
> If so then this change is not necessary but previously this was the issue with ceph-volume tox test. Does every VM for test is the same?

@jansobczak not sure about that. I would need to check all jenkins instances.
That being said, I could make the test pass on a machine where `/dev/cdrom` doesn't exist with something like following:

```python
    @patch('ceph_volume.util.device.os.readlink')
    @patch('ceph_volume.util.device.os.path.islink')
    def test_reject_readonly_device(self, m_os_path_islink, m_os_readlink, device_info):
        m_os_path_islink.return_value = True
        m_os_readlink.return_value = '/dev/cdrom'
        data = {"/dev/cdrom": {"ro": 1}}
        lsblk = {"TYPE": "disk"}
        device_info(devices=data,lsblk=lsblk)
        disk = device.Device("/dev/cdrom")
        assert not disk.available
```

Reply:
```console
[ceph: root@cv-pr-38215 ceph_volume]# ls /dev/cdrom
ls: cannot access '/dev/cdrom': No such file or directory
[ceph: root@cv-pr-38215 ceph_volume]# pytest --pdb -v tests/util/test_device.py::TestDevice::test_reject_readonly_device
========================================================================================================================================================================================================================================== test session starts ===========================================================================================================================================================================================================================================
platform linux -- Python 3.6.8, pytest-6.2.4, py-1.10.0, pluggy-0.13.1 -- /usr/bin/python3.6
cachedir: .pytest_cache
rootdir: /usr/lib/python3.6/site-packages/ceph_volume
collected 1 item

tests/util/test_device.py::TestDevice::test_reject_readonly_device PASSED                                                                                                                                                                                                                                                                                                                                                                                                                          [100%]
```

Reply:
I'm not convinced this check should check `/dev/cdrom` at all @guits.
If we take a look at the test for `non_readonly_device` which looks like 
```    def test_accept_non_readonly_device(self, device_info):
        data = {"/dev/sda": {"ro": 0, "size": 5368709120}}
        lsblk = {"TYPE": "disk"}
        device_info(devices=data,lsblk=lsblk)
        disk = device.Device("/dev/sda")
        assert disk.available
```
Then the test for a read-only device should differ with 
```diff
- data = {"/dev/sda": {"ro": 0, "size": 5368709120}}
+ data = {"/dev/sda": {"ro": 1, "size": 5368709120}}
```
The disk of type `rom` is filtered at disk.py: https://github.com/ceph/ceph/blob/master/src/ceph-volume/ceph_volume/util/disk.py#L755

If a VM's are not the same then I guess the only way is to ditch testing which require presence of  `/dev/cdrom` or make VMs configuration more consistent. 

Your example gave me an idea to create those tests:
```python
    @patch('ceph_volume.util.device.os.readlink')
    @patch('ceph_volume.util.device.os.path.islink')
    def test_accept_symlink_to_device(self, m_os_path_islink, m_os_readlink, device_info):
        m_os_path_islink.return_value = True
        m_os_readlink.return_value = '/dev/sdb'
        data = {"/dev/sdb": {"ro": 0, "size": 5368709120}}
        lsblk = {"TYPE": "disk"}
        device_info(devices=data,lsblk=lsblk)
        disk = device.Device("/dev/test_symlink")
        print(disk)
        print(disk.sys_api)
        assert disk.available

    @patch('ceph_volume.util.device.os.readlink')
    @patch('ceph_volume.util.device.os.path.islink')
    def test_reject_symlink_to_device_mapper(self, m_os_path_islink, m_os_readlink, device_info):
        m_os_path_islink.return_value = True
        m_os_readlink.return_value = '/dev/dm-0'
        data = {"/dev/mapper/mpatha": {"ro": 0, "size": 5368709120}}
        lsblk = {"TYPE": "disk"}
        device_info(devices=data,lsblk=lsblk)
        disk = device.Device("/dev/mapper/mpatha")
        assert disk.available
```

Reply:
> I'm not convinced this check should check `/dev/cdrom` at all @guits.

me neither, I'm not sure why it was written like this but perhaps I'm missing a detail 🤷‍♂️ 

> If we take a look at the test for `non_readonly_device` which looks like
> 
> ```
>         data = {"/dev/sda": {"ro": 0, "size": 5368709120}}
>         lsblk = {"TYPE": "disk"}
>         device_info(devices=data,lsblk=lsblk)
>         disk = device.Device("/dev/sda")
>         assert disk.available
> ```
> 
> Then the test for a read-only device should differ with
> 
> ```diff
> - data = {"/dev/sda": {"ro": 0, "size": 5368709120}}
> + data = {"/dev/sda": {"ro": 1, "size": 5368709120}}
> ```
> 
> The disk of type `rom` is filtered at disk.py: https://github.com/ceph/ceph/blob/master/src/ceph-volume/ceph_volume/util/disk.py#L755
> 
> If a VM's are not the same then I guess the only way is to ditch testing which require presence of `/dev/cdrom` or make VMs configuration more consistent.
> 
> Your example gave me an idea to create those tests:
> 
> ```python
>     @patch('ceph_volume.util.device.os.readlink')
>     @patch('ceph_volume.util.device.os.path.islink')
>     def test_accept_symlink_to_device(self, m_os_path_islink, m_os_readlink, device_info):
>         m_os_path_islink.return_value = True
>         m_os_readlink.return_value = '/dev/sdb'
>         data = {"/dev/sdb": {"ro": 0, "size": 5368709120}}
>         lsblk = {"TYPE": "disk"}
>         device_info(devices=data,lsblk=lsblk)
>         disk = device.Device("/dev/test_symlink")
>         print(disk)
>         print(disk.sys_api)
>         assert disk.available
> 
>     @patch('ceph_volume.util.device.os.readlink')
>     @patch('ceph_volume.util.device.os.path.islink')
>     def test_reject_symlink_to_device_mapper(self, m_os_path_islink, m_os_readlink, device_info):
>         m_os_path_islink.return_value = True
>         m_os_readlink.return_value = '/dev/dm-0'
>         data = {"/dev/mapper/mpatha": {"ro": 0, "size": 5368709120}}
>         lsblk = {"TYPE": "disk"}
>         device_info(devices=data,lsblk=lsblk)
>         disk = device.Device("/dev/mapper/mpatha")
>         assert disk.available
> ```

do you mind updating your branch with all those suggested changes?

(sorry for the delay by the way, I'm just returning from PTO today..)

Reply:
Apologies for late update